### PR TITLE
Contacts

### PR DESCRIPTION
**Pull request by jhuebel**
OK, so I'm not 100% satisfied with this but felt like it was time to ship it. Release early, release often.

Anyway, I deviated from the way Flarepoint was doing things in a few ways:

First, my changes avoid the use of the Repository pattern where possible. Unless someone sees a compelling reason to stick with repos, I feel like they just add an extra layer of abstraction that doubles work. Personally, I didn't consider the ability to change database backends as a strong enough reason to keep using repos. The unfortunate effect of this is that we now have a mix of repos and direct Eloquent implementation. My intention at some point in the future is to rewrite all the repo code to be plain old Eloquent. In the meantime, it'll be a little messy.

Second, contact-related pages are laid out with responsive design in mind, although that's not fully implemented. I'll work more on that soon.

Third, three new contact permissions were created: contact-create, contact-update and contact-delete. It occurs to me that I may not have implemented these new permissions in the best way (I just added them to PermissionsTableSeeder), since they won't be created if you're upgrading. So I'll have to commit an update to this pull request later to fix that.

Lastly, I setup a four-level hierarchy for user roles: Administrator, Manager, Sales Representative and Employee. The idea is that Managers can modify all client data, but can't create users. Sales Reps create and edit, but can't delete. And Employees are read-only. These roles are only created for a new install at the moment. No effort was made to migrate your existing roles. I probably should have made this change in another branch, but it's too late now. 😁 

I would recommend against trying to upgrade an existing Flarepoint install at the moment. Especially if you have data you care about.

### CODE REVIEW

**Comment by jhuebel** [comment]
OK, just commited the change so people can upgrade to the new contact permissions. Basically, had to move the new permissions from the PermissionsTableSeeder to a migration.

**Comment by jhuebel** [comment]
Also, the large number of changes are primarily due to the PSR2 and Symfony formatting I applied to all the files using PHP-CS-Fixer. In retrospect, I should have waited on that until later or only reformatted files as I touched them. But the consistent formatting helped me understand the code better.

**Comment by jhuebel** [comment]
Upgrading:

1. Pull the changes.
2. Run: composer install
3. Run: php artisan migrate
4. In production, make sure APP_DEBUG is set to false in your .env file. Otherwise, Debugbar will show up.
5. Login to Flarepoint as the admin user.
6. Go to Settings -> Overall Settings.
7. Enable the "Create contact", "Update contact" and "Delete contact" permissions for each role that should have those permissions.

After step 2, you could also run "npm audit fix", since there are a number of vulnerable node modules. But I didn't include that in the upgrade instructions above because it isn't strictly required.
That should do the trick.

**Comment by jhuebel** [comment]
There's a link in the latest commit (changing the README) that points to a Wiki page about Upgrading. Obviously that doesn't exist yet, but can basically be the upgrade instructions I provided earlier.

File: app/Models/Client.php [hunk 1]
```diff
@@ -39,6 +63,13 @@ public function leads()
             ->orderBy('created_at', 'desc');
     }
 
+    public function contacts()
```
Comment (Bottelet) [inline app/Models/Client.php]: This function sorts, I don't think the sorting should be done there but where it's used instead, but a small thing.

File: composer.json [hunk 2]
```diff
@@ -25,7 +26,9 @@
         "pusher/pusher-php-server": "^2.3",
         "zizaco/entrust": "1.7.*",
         "laravel/tinker": "^1.0",
-        "laravel/dusk": "^1.0"
+        "laravel/dusk": "^1.0",
+        "doctrine/dbal": "^2.9",
```
Comment (Bottelet) [inline composer.json]: Why are we adding doctrine? Where is it even used?

File: app/Models/Client.php (hunk 1 above)
Comment (jhuebel) [reply app/Models/Client.php]: Sorting seemed to be the norm in the other relationships that already existed, so I honored that in my contacts relationship as well. I think I just copied the tasks relationship and changed the class it used (from Tasks::class to Contacts::class).

File: composer.json (hunk 2 above)
Comment (jhuebel) [reply composer.json]: Doctrine DBAL allows you to modify existing database columns using a migration.

See here: https://laravel.com/docs/5.4/migrations#modifying-columns

I used DBAL's change() method in 2019_05_05_233153_change_contact_field_name.php.

**Comment by jhuebel** [comment]
> 1. The action button for creating new contacts, tasks etc on the client view is not working.

You're talking about the "New Client" button on right side of the Contacts table header? Worked for me.

> 2. First time creating a client, should it not create a contact by default with the information provided from the client form?

Honestly, I hadn't considered that as an option. I left the contact information alone in the client record to avoid the loss of data. But at some point, I thought that we would remove the contact information from the client record entirely (maybe in a future version). But perhaps hiding those fields (so data is still preserved temporarily) would be the way to go.

> 3. the action view on the client, could be cool if it used the same table as the others to keep the same design.

Do you mean changing the client.show view to be laid out like I did for the contacts.show view? That was my intention eventually. But I wanted to give you a chance to see the direction I was going with it (and say "no") before I started changing the look-and-feel of anything that already existed. Granted, I made some tweaks to some of the list views (the combined "Actions" column). But I didn't want to make any significant alterations to your layouts until we had a chance to discuss them.

**Comment by jhuebel** [comment]
I've given your second question (about creating a contact record) some additional thought.

Here's what I suggest:

Now that clients and contacts are separate, the contact information in the client record should be used for the general business address (or move to separate shipping and billing addresses), main business phone/FAX numbers and the main business email address (for instance, the Accounts Payable email address).

The contacts should be used for individual contacts whose phone numbers, email addresses and branch addresses you need to keep track of.

**Comment by Bottelet** [comment]
@jhuebel on the client view? under each table there is a create button action does that work for you? for me they all link to create client. The deisng i mostly mean it being green, which stands out a bit to much, think it should have the same color and border as the others, which are blue.

**Comment by jhuebel** [comment]
> @jhuebel on the client view? under each table there is a create button action does that work for you? for me they all link to create client. The deisng i mostly mean it being green, which stands out a bit to much, think it should have the same color and border as the others, which are blue.

I'm not seeing that here. Each button appears to be working correctly for me. I'm assuming you mean these buttons in the bottom right of this screen snippet?

image

If so, each one works perfectly for me. Each button does append `?client=<#>` to the end of the URL. That's to tell each form to associate the contact, lead or task with that client.

File: app/Models/Client.php (hunk 1 above)
Comment (Bottelet) [reply app/Models/Client.php]: Fair enough :) No problems with this, yes there are some questionable choices in the code :)

**Comment by Bottelet** [comment]
Yes exactly those :) Hmm weird, I'll double check tommrow.

**Comment by jhuebel** [comment]
I'd like to consider this PR as done, even though we've got a mix of responsive design in my contacts pages and the original layout in the other pages. But I'm moving to a new branch for the overhaul of the interface to use responsive layout/design.

**Pull request merged**